yellow-paper/docs/gas-and-fees/gas-metering-and-fees.md

```diff
@@ -0,0 +1,409 @@
+- `max_fee_per_da_gas`: the maximum price the user is willing to pay per unit of DA gas in `$AZT`
+- `compute_gas_limit`: the maximum amount of compute gas the user is willing to pay for
+- `max_fee_per_compute_gas`: the maximum price the user is willing to pay per unit of compute gas in `$AZT`
+- `inclusion_fee`: the flat fee the user is willing to pay for transaction inclusion in `$AZT`
+
+
+# Sequencer checks
+
+:::warning
+TODO add checks that rational sequencers will perform to reduce the risk of wasted work.
+:::
+
+
+# Tracking in kernels
+
+### GasUsed
+
+We have a `GasUsed` that contains:
+
+- `da_gas_used`
+- `compute_gas_used`
+
+### GasLimits
+
+We have a `GasLimits` that contains:
+
+- `da_gas_limit`
+- `max_fee_per_da_gas`
+- `compute_gas_limit`
+- `max_fee_per_compute_gas`
+- `inclusion_fee`
+
+### CallContext
+
+- `gas_limits`
+- `transaction_fee`
+  - Only non-zero during teardown
+
+### PrivateKernelPublicInputs
+
+- `gas_limits`
+- `end`: `PrivateAccumulatedData`
+
+### PrivateAccumulatedData
+
+- `gas_used`
+
+### PublicKernelPublicInputs
+
+- `gas_limits`
+- `end`: `PublicAccumulatedData`
+- `end_non_revertible`: `PublicAccumulatedData`
+
+### PublicAccumulatedData
+
+- `gas_used`
+
+### PublicCircuitPublicInputs
+
+- `fee_paid`
+  - false unless the `pay_fee` function was called in teardown
+
+### KernelCircuitPublicInputs
+
+- `gas_limits`
+- `transaction_fee`
+  - Summed during public kernel tail
+  - Zero-ed during public kernel teardown if `fee_paid` is false
+- `end`: `AccumulatedData`
+
+### AccumulatedData
+
+- `gas_used`
+
+### GlobalVariables
+
+- `fee_per_da_gas`
+- `fee_per_compute_gas`
+
+
+# Pure private transactions
+
+Purely private transactions cannot pay a fee, and thus the `transaction_fee` is set to 0.
+
+However, they still consume gas, and thus the `da_gas_used` and `compute_gas_used` must be tracked.
+
+```
+kernel_public_inputs.gas_limits := private_kernel_public_inputs.gas_limits
+
+kernel_public_inputs.end.gas_used.da_gas_used :=
+    FIXED_DA_GAS
+    + private_kernel_public_inputs.end.gas_used.da_gas_used
+
+kernel_public_inputs.end.compute_gas_used :=
+    private_kernel_public_inputs.end.compute_gas_used
+
+# make assertions that gas limits are not exceeded
+
+# purely private transactions cannot pay any fee
+kernel_public_inputs.transaction_fee := 0
+```
+
+# Transaction with public component
+
+If a transaction has a public component, `end_non_revertible.gas_used.da_gas_used` is initialized to `FIXED_DA_GAS` in the `PrivateKernelTailToPublic` kernel.
+
+### Interaction with the public VM
+
+The public VM retains the current amount of `daGasLeft` in its [MachineState](../public-vm/state.md#machine-state).
+
+This amount must be initialized for each public kernel iteration as:
+
+```
+daGasLeft := public_kernel_public_inputs.da_gas_limit 
+             - public_kernel_public_inputs.end_non_revertible.da_gas_used
+             - public_kernel_public_inputs.end.da_gas_used
+```
```

Comment:
IIUC this means we'll have a single global gas limit for all enqueued public calls, which doesn't match the API we're offering the user for public calls made within the AVM, where the user can specify individual gas limits for each call. This is useful (in EVM-land at least) to be able to limit how much gas you're spending when calling a potentially untrusted contract, and being able to "catch" the out of gas and continue execution. 

However, as I'm writing this, I've noticed you cannot "catch" an out of gas from an _enqueued_ function, since it's called asynchronously. So an oog in an enqueued function would always revert the whole tx, and there is not much of a point in setting differentiated gas limits for them.

But does it make sense to have differentiated gas limits for non-revertible (setup) and revertible (app) public functions? Would they help in protecting the sequencer, by letting them know in advance how much work they must do (ie compute gas) until they can secure their fee? Or would they help in protecting the user somehow?